### PR TITLE
Turn `CallPath` into a wrapper type

### LINKED ISSUE #2018: Allow the use of type `Self` for struct expressions
This does work in some cases, but something like:
```rust
impl<T> RawVec<T> {
    /// Create a new `RawVec` with zero capacity.
    fn new() -> Self { // Returning `Self` here is fine
        Self { // <- Problem is here.. Need to replace `Self` with `RawVec` for this to work.
            ptr: alloc(0),
            cap: 0,
        }
    }
}
```
does not work and gives `Could not find symbol "Self" in this scope`

### PR DESCRIPTION
**Pull request by emilyaherbert**
This PR introduces `CallPath<T>` where `suffix: T`. `CallPath<T>` is used to represent `CallPath<Ident>` and `CallPath<(TypeInfo, Span)>`, representing two different cases in which the suffix of the call path is either an ident pointing towards a declaration, or the name of a type.

- Blocked by #2201
- Closes #2018

### CODE REVIEW

**Comment by sezna** [comment]
I know you're already kinda into this, but what about a `Spanned<T>` type to be consistent with the parser? E.g. `Spanned<TypeInfo>` is a `TypeInfo` with a span.

**Comment by emilyaherbert** [comment]
> I know you're already kinda into this, but what about a Spanned<T> type to be consistent with the parser? E.g. Spanned<TypeInfo> is a TypeInfo with a span.

Wait @sezna, I'm not seeing how this would work... `Spanned` is a trait with the method `span`, it would not add any `span` field to `TypeInfo`.

File: sway-core/src/parse_tree/call_path.rs [hunk 1]
```diff
@@ -5,16 +5,17 @@ use crate::Ident;
 use sway_types::{span::Span, Spanned};
 
 /// in the expression `a::b::c()`, `a` and `b` are the prefixes and `c` is the suffix.
+/// `c` can be any type `T`, but in practice `c` is either an `Ident` or a `TypeInfo`.
 #[derive(Debug, Clone, Eq, PartialEq, Hash)]
-pub struct CallPath {
+pub struct CallPath<T> {
```
Comment (mitchmindtree) [inline sway-core/src/parse_tree/call_path.rs:10]: I wonder if we could specify `Ident` as a default type here in the case that one isn't specified? E.g.

```suggestion
pub struct CallPath<T = Ident> {
```

This might help to easy the common case and cut down the diff a little?

Comment (emilyaherbert) [reply sway-core/src/parse_tree/call_path.rs:10]: Good idea!

File: sway-core/src/parse_tree/call_path.rs [hunk 2]
```diff
@@ -24,19 +25,25 @@ impl std::convert::From<Ident> for CallPath {
...
+    T: fmt::Display,
+{
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         let mut buf = String::new();
         for prefix in self.prefixes.iter() {
             buf.push_str(prefix.as_str());
             buf.push_str("::");
         }
-        buf.push_str(self.suffix.as_str());
+        buf.push_str(&self.suffix.to_string());
```
Comment (mitchmindtree) [inline sway-core/src/parse_tree/call_path.rs:38]: **_Hot tip_**: Not super important, but if `suffix` is a `T` where `T: Display`, you can do:

```rust
write!(&mut buf, "{}", self.suffix);
```
which avoids the extra `String` alloc from `to_string()`.

Alternatively, this whole `impl` should probably write directly into the `Formatter` with `write!(f, ...)`, but a change like that could probably be left for a separate PR.

File: sway-core/src/semantic_analysis/namespace/root.rs [hunk 3]
```diff
@@ -64,34 +64,58 @@ impl Root {
...
+        self.resolve_type(
+            type_id,
+            span,
+            enforce_type_arguments,
+            type_info_prefix,
+            mod_path,
+        )
+    }
+
     pub(crate) fn resolve_type(
```
Comment (mitchmindtree) [inline sway-core/src/semantic_analysis/namespace/root.rs:86]: Perhaps this could be a good time to add some docs for `resolve_type` and `resolve_type_with_self`, particularly as we're adding a new `type_info_prefix` arg?

**Comment by otrho** [comment]
Looks like there's some conflicts with the new IR tests being moved.  I found trying to resolve these could be a pain, I can help out if you like.

File: sway-core/src/semantic_analysis/namespace/root.rs (hunk 3 above)
Comment (emilyaherbert) [reply sway-core/src/semantic_analysis/namespace/root.rs:86]: I added this to this issue: https://github.com/FuelLabs/sway/issues/1673

**Comment by emilyaherbert** [comment]
@otrho the IR tests seem to pass right out of the box---which makes sense as there should be no behavior change other than introducing a new feature.

File: sway-core/tests/sway_to_ir/mutable_struct.ir [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    fn main() -> u64, !1 {
+        local mut ptr { u64, u64 } record
+
+        entry:
+        v0 = const { u64, u64 } { u64 undef, u64 undef }, !2
+        v1 = const u64 40, !3
+        v2 = insert_value v0, { u64, u64 }, v1, 0, !2
+        v3 = const u64 2, !4
+        v4 = insert_value v2, { u64, u64 }, v3, 1, !2
+        v5 = get_ptr mut ptr { u64, u64 } record, ptr { u64, u64 }, 0, !5
```
Comment (otrho) [inline sway-core/tests/sway_to_ir/mutable_struct.ir:11]: It looks like you're re-instating these files when they should be deleted?  Or am I reading this wrongly?  The `sway_to_ir` directory shouldn't exist on `master` any more.

File: sway-core/tests/sway_to_ir/mutable_struct.ir [hunk 5]
```diff
@@ -0,0 +1,31 @@
+script {
```
Comment (mohammadfawaz) [inline sway-core/tests/sway_to_ir/mutable_struct.ir:1]: Toby moved all IR tests to https://github.com/FuelLabs/sway/tree/master/test/src/ir_generation and changed them to use `FileCheck`.

File: sway-lib-std/src/vec.sw [hunk 6]
```diff
@@ -14,7 +14,7 @@ struct RawVec<T> {
 impl<T> RawVec<T> {
     /// Create a new `RawVec` with zero capacity.
     fn new() -> Self {
-        RawVec {
+        Self {
```
Comment (mohammadfawaz) [inline sway-lib-std/src/vec.sw:17]: Nice 😄

**Review by mohammadfawaz — COMMENTED** [review]
Looks great! Just the comment about the IR tests.

File: sway-core/tests/sway_to_ir/mutable_struct.ir (hunk 4 above)
Comment (emilyaherbert) [reply sway-core/tests/sway_to_ir/mutable_struct.ir:11]: Good eye---I didn't realize that the merge added these files back in.

File: sway-core/tests/sway_to_ir/mutable_struct.ir (hunk 5 above)
[the remark this answers is not in the document]
Comment (emilyaherbert) [reply sway-core/tests/sway_to_ir/mutable_struct.ir:1]: Yes, this was a snafu with the merge.

**Review by mohammadfawaz — APPROVED** [review]

File: sway-core/src/semantic_analysis/node_dependencies.rs [hunk 7]
```diff
@@ -574,7 +574,10 @@ impl Dependencies {
                 type_arguments,
             } => {
                 self.deps.insert(DependentSymbol::Symbol(name.clone()));
-                self.gather_from_type_arguments(type_arguments)
+                match type_arguments {
+                    Some(type_arguments) => self.gather_from_type_arguments(type_arguments),
+                    None => self,
+                }
```
Comment (otrho) [inline sway-core/src/semantic_analysis/node_dependencies.rs:580]:
```suggestion
                type_arguments
                    .map(|type_arguments| self.gather_from_type_arguments(type_arguments))
                    .unwrap_or(self)
```

**Review by otrho — APPROVED** [review]
